lib/astc_encode.cpp

```diff
@@ -0,0 +1,963 @@
+                           work->context, work->image, &work->swizzle,
+                           work->data_out, work->data_len, threadId);
+
+    // This is a racy update, so which error gets returned is a random, but it
+    // will reliably report an error if an error occurs
+    if (error != ASTCENC_SUCCESS) {
+        work->error = error;
+    }
+}
+
+/**
+ * @brief Worker thread helper payload for launchThreads.
+ */
+struct LaunchDesc {
+    /** The native thread handle. */
+    pthread_t threadHandle;
+    /** The total number of threads in the thread pool. */
+    int threadCount;
+    /** The thread index in the thread pool. */
+    int threadId;
+    /** The user thread function to execute. */
+    void (*func)(int, int, void*);
+    /** The user thread payload. */
+    void* payload;
+};
+
+/**
+ * @brief Helper function to translate thread entry points.
+ *
+ * Convert a (void*) thread entry to an (int, void*) thread entry, where the
+ * integer contains the thread ID in the thread pool.
+ *
+ * @param p The thread launch helper payload.
+ */
+static void*
+launchThreadsHelper(void *p) {
+    LaunchDesc* ltd = (LaunchDesc*)p;
+    ltd->func(ltd->threadCount, ltd->threadId, ltd->payload);
+    return nullptr;
+}
+
+/* Public function, see header file for detailed documentation */
+static void
+launchThreads(int threadCount, void (*func)(int, int, void*), void *payload) {
+    // Directly execute single threaded workloads on this thread
+    if (threadCount <= 1) {
+        func(1, 0, payload);
+        return;
+    }
+
+    // Otherwise spawn worker threads
+    LaunchDesc *threadDescs = new LaunchDesc[threadCount];
+    for (int i = 0; i < threadCount; i++) {
+        threadDescs[i].threadCount = threadCount;
+        threadDescs[i].threadId = i;
+        threadDescs[i].payload = payload;
+        threadDescs[i].func = func;
+
+        pthread_create(&(threadDescs[i].threadHandle), nullptr,
+                       launchThreadsHelper, (void*)&(threadDescs[i]));
+    }
+
+    // ... and then wait for them to complete
+    for (int i = 0; i < threadCount; i++) {
+        pthread_join(threadDescs[i].threadHandle, nullptr);
+    }
+
+    delete[] threadDescs;
+}
+
+/**
+ * @memberof ktxTexture
+ * @ingroup writer
+ * @~English
+ * @brief Encode and compress a ktx texture with uncompressed images to astc.
+ *
+ * The images are encoded to astc block-compressed format. The encoded images
+ * replace the original images and the texture's fields including the DFD are
+ * modified to reflect the new state.
+ *
+ * Such textures can be directly uploaded to a GPU via a graphics API.
+ *
+ * @param[in]   This   pointer to the ktxTexture object of interest.
+ * @param[in]   params pointer to astc params object.
+ *
+ * @return      KTX_SUCCESS on success, other KTX_* enum values on error.
+ *
+ * @exception KTX_INVALID_OPERATION
+ *                              The texture is already supercompressed.
```

Comment:
The texture's images are supercompressed.

Please make the same change in `ktxTexture2_CompressBasisEx` if necessary.

Reply:
Fixed in https://github.com/KhronosGroup/KTX-Software/pull/433/commits/c01d05d4cfa0353f848c18051d90ba94862115b5